### PR TITLE
Improve Normalization Schedule Performance

### PR DESCRIPTION
**Pull request by rdspring1**
Incorporate Deyu's suggestions into the Magic Scheduler

1. Use Outer-Split to coalesce memory access
2. Enable Fast-Math flag
3. Cast double scalars to float in binary operations
4. Implement Cache-Fork - Avoid global read for intermediate fusion outputs

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/fusion.cpp [hunk 1]
```diff
@@ -240,6 +240,25 @@ void Fusion::removeOutput(Val* output) {
...
+
+    if (replacement->getValType().value() == ValType::TensorView) {
+      replacement->setIsFusionOutput(true);
+      replacement->as<TensorView>()->setMemoryType(MemoryType::Global);
+    }
+    if (output->getValType().value() == ValType::TensorView) {
+      output->setIsFusionOutput(false);
+      output->as<TensorView>()->setMemoryType(MemoryType::Local);
+    }
+  }
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:259]: Need to call `resetTvUses()`

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 2]
```diff
@@ -966,16 +967,29 @@ std::vector<TensorView*> findTensorViewsToDuplicate(
   return duplicate_tv;
 }
 
+bool canComputeAtInline(Fusion* fusion, TensorView* tv) {
+  auto uses = fusion->unordered_uses(tv);
+  if (uses.size() == 1) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp:971]: CC @naoyam would be interesting to have a less conservative check for this.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 3]
```diff
@@ -1134,6 +1148,15 @@ void scheduleNormalization(
 
   organizeAxes(reduction_tv, all_tv);
 
+  // For intermediate outputs, apply cache_fork
+  for (const auto output : fusion->outputs()) {
+    if (!fusion->unordered_uses(output).empty()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: nit: use `output->uses`

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 4]
```diff
@@ -692,6 +692,55 @@ TensorView* TensorView::cache_before() {
...
 
+TensorView* TensorView::cache_fork() {
+  FusionGuard fg(fusion());
+
+  // Before: [Expr] -> This TV (Global Output) -> [Usage Expr]
+  // After:  [Expr] -> This TV (Local) -> [Usage Expr] > Next TV
+  //                            (Fork) -> [Set Expr]   -> New TV (Global Output)
+
+  TORCH_CHECK(
+      fusion()->hasOutput(this) && !fusion()->unordered_uses(this).empty(),
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: nit: use `this->uses()`

**Review by csarofeen — CHANGES_REQUESTED** [review]
Minor change needed. otherwise LGTM

File: torch/csrc/jit/codegen/cuda/fusion.cpp (hunk 1 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:259]: :+1:

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 5]
```diff
@@ -966,16 +967,29 @@ std::vector<TensorView*> findTensorViewsToDuplicate(
   return duplicate_tv;
 }
 
+bool canComputeAtInline(Fusion* fusion, TensorView* tv) {
+  auto uses = fusion->unordered_uses(tv);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: nit: tv->uses

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 6]
```diff
@@ -692,6 +692,55 @@ TensorView* TensorView::cache_before() {
...
+
+  // Set the computeAt for this forked TensorView
+  // to the Fusion outputs without any uses
+  if (hasComputeAt()) {
+    auto this_ca_pos = getThisComputeAtAxis();
+    auto rel_ca_pos = getRelativeComputeAtAxis();
+
+    for (Val* out : fusion()->outputs()) {
+      if (out->getValType() == ValType::TensorView) {
+        if (fusion()->unordered_uses(out).empty()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: nit: out->uses

**Review by csarofeen — DISMISSED** [review]

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 7]
```diff
@@ -343,6 +343,42 @@ class CudaKernelGenerator : private kir::IrVisitor {
     return expr.str();
   }
 
+  // If one argument is a tensorview and the other is a scalar, make sure we
+  // cast the scalar to the tensorview type
+  std::string scalarCast(kir::Val* lhs, kir::Val* rhs) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/codegen.cpp:348]: @shmsong could you please review this section to make sure it's consistent with pytorch type promotion?

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 8]
```diff
@@ -364,7 +364,7 @@ NvrtcFunction nvrtcCompile(
   const std::string compute = "--gpu-architecture=compute_" +
       std::to_string(major) + std::to_string(minor);
   std::vector<const char*> args = {
-      "--std=c++14", compute.c_str(), "-default-device"};
+      "--std=c++14", "--use_fast_math", compute.c_str(), "-default-device"};
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp:367]: @jjsjann123 @kevinstephano any opinions on turning on fast math by default?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:367]: No objection.

Just curious: "--use_fast_math" only changes flush denorm to zero. Is that a game changer in terms of perf?

Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:367]: According to Deyu's Softmax experiments:
1. Baseline: ~200us
2. Outer Split - Coalesce Gmem: ~110us
3. Add fast-math compile flag: ~43us

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 7 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:348]: This part is consistent when the `TensorView` has floating point. https://github.com/csarofeen/pytorch/blob/20_12_3_devel/aten/src/ATen/native/TypeProperties.cpp#L59-L73

@rdspring1 , just wanted to mention that when `TensorView` is integer and `Scalar` is floating point eager mode will cast the `TensorView` to floating point. Not sure if this case is relevant/reachable here. 

Our duplication of the type combine logic is here if it helps. 
https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/arith.cpp#L246-L259

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp (hunk 8 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:367]: At least 10-15% improvement according to softmax_dropout benchmarks when adding fast-math flag

**Review by csarofeen — APPROVED** [review]
